components/candle/masteryService.ts

```diff
@@ -0,0 +1,198 @@
+import {
+    MasteryData,
+    MasteryDataTemplate,
+    MasteryDrop,
+    MasteryPackage,
+} from "../../components/types/mastery"
+import {
+    CompletionData,
+    GameVersion,
+    Unlockable,
+} from "../../components/types/types"
+import { xpRequiredForLevel } from "../../components/utils"
+
+export class MasteryService {
+    // @ts-expect-error Unused
+    private controller: Controller
+    private masteryData: Map<string, MasteryPackage> = new Map()
+
+    constructor(controller: Controller) {
+        this.controller = controller
+    }
+
+    registerMasteryData(masteryPackage: MasteryPackage) {
+        this.masteryData.set(masteryPackage.Id, masteryPackage)
+    }
+
+    getMasteryDataForDestination(
+        locationParentId: string,
+        gameVersion: GameVersion,
+        userId: string,
+    ): MasteryData[] {
+        return this.getMasteryData(locationParentId, gameVersion, userId)
+    }
+
+    getMasteryDataForLocation(
+        locationId: string,
+        gameVersion: GameVersion,
+        userId: string,
+    ): MasteryDataTemplate {
+        const subLocation: Unlockable = getSubLocationByName(
+            locationId,
+            gameVersion,
+        )
+
+        const masteryDataTemplate: MasteryDataTemplate =
+            getConfig<MasteryDataTemplate>(
+                "MasteryDataForLocationTemplate",
+                true,
+            )
+
+        const masteryData = this.getMasteryData(
+            subLocation.Properties.ParentLocation,
+            gameVersion,
+            userId,
+        )
+
+        return <MasteryDataTemplate>{
+            template: masteryDataTemplate,
+            data: {
+                MasteryData: masteryData,
+            },
+        }
+    }
+
+    getCompletionData(
+        locationParentId: string,
+        gameVersion: GameVersion,
+        userId: string,
+    ): CompletionData {
+        if (!this.masteryData.has(locationParentId)) {
+            return undefined
+        }
+
+        //Get the mastery data
+        const masteryData: MasteryPackage =
+            this.masteryData.get(locationParentId)
+
+        //Get the user profile
+        const userProfile = getUserData(userId, gameVersion)
+
+        //Gather all required data
+        const lowerCaseLocationParentId = locationParentId.toLowerCase()
+
+        if (
+            !userProfile.Extensions.progression.Locations[
+                lowerCaseLocationParentId
+            ]
+        ) {
+            userProfile.Extensions.progression.Locations[
+                lowerCaseLocationParentId
+            ] = {
+                Xp: 0,
+                Level: 1,
+            }
+        }
+
+        const locationData =
+            userProfile.Extensions.progression.Locations[
+                lowerCaseLocationParentId
+            ]
+
+        const maxLevel =
+            masteryData.MaxLevel ||
+            masteryData.Drops.reduce((currentMax, drop) => {
```

Comment:
Do you think having this just default to 20 would be more beneficial? As there are only 2 exceptions (Evergreen and Carpathian being 25-100 and 5 respectively).

Reply:
Yes, I would go with a default of 20 and the let the JSON tell you if it's supposed to be something else.